test/unit-test/test_memory_functional.c

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Shift Cryptosecurity AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <setjmp.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <cmocka.h>
+
+#include <memory/memory.h>
+
+#define CHUNK_SIZE (8 * 1024)
+#define NUM_CHUNKS (3)
+
+// init NUM_CHUNKS with 0xFF.
+__extension__ static uint8_t _memory[] = {[0 ...(CHUNK_SIZE * NUM_CHUNKS) - 1] = 0xFF};
+
+bool __wrap_memory_write_chunk_mock(uint32_t chunk_num, uint8_t* chunk)
+{
+    memcpy(&_memory[chunk_num * CHUNK_SIZE], chunk, CHUNK_SIZE);
+    return true;
+}
+
+void __wrap_memory_read_chunk_mock(uint32_t chunk_num, uint8_t* chunk_out)
+{
+    memcpy(chunk_out, &_memory[chunk_num * CHUNK_SIZE], CHUNK_SIZE);
+}
+
+static void _test_memory_multisig(void** state)
+{
+    const uint8_t hashes[][32] = {
+        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+        "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
+        "cccccccccccccccccccccccccccccccc",
+        "dddddddddddddddddddddddddddddddd",
+        "eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
+        "ffffffffffffffffffffffffffffffff",
+    };
+    const char* names[] = {
+        "name1",
+        "name2",
+        "name3",
+        "name4",
+        "name5",
+        "name6",
+    };
+
+    char name[31] = {0};
+    assert_false(memory_multisig_get_by_hash(hashes[0], name));
+
+    // set
+    assert_int_equal(
+        MEMORY_MULTISIG_SET_BY_HASH_RESULT_OK, memory_multisig_set_by_hash(hashes[0], names[0]));
+    assert_int_equal(
+        MEMORY_MULTISIG_SET_BY_HASH_RESULT_OK, memory_multisig_set_by_hash(hashes[1], names[1]));
+    // overwrite with the same is possible
+    assert_int_equal(
+        MEMORY_MULTISIG_SET_BY_HASH_RESULT_OK, memory_multisig_set_by_hash(hashes[1], names[1]));
+
+    // get
+    assert_true(memory_multisig_get_by_hash(hashes[0], NULL));
+    assert_true(memory_multisig_get_by_hash(hashes[0], name));
+    assert_string_equal(name, names[0]);
+    assert_true(memory_multisig_get_by_hash(hashes[1], name));
+    assert_string_equal(name, names[1]);
+    // rename
+    const char* name0_renamed = "name 1 renamed";
+    assert_int_equal(
+        MEMORY_MULTISIG_SET_BY_HASH_RESULT_OK,
+        memory_multisig_set_by_hash(hashes[0], name0_renamed));
+    assert_true(memory_multisig_get_by_hash(hashes[0], name));
+    assert_string_equal(name, name0_renamed);
+
+    // rename to a name which already exists fails (duplicate name).
```

Comment:
Could you also check that the old data still is there after a failed write?

Reply:
Done.